cms/cms_toolbars.py

```diff
@@ -86,6 +88,107 @@ def add_wizard_button(self):
+            return
+
+        # Add edit button when in content mode.
+        if self.toolbar.content_mode_active:
+            self.add_edit_button()
+        # Add preview button when in edit mode.
+        if self.toolbar.edit_mode_active:
+            self.add_preview_button()
+        # Always add structure mode.
+        self.add_structure_mode()
+
+    def init_placeholders(self):
+        request = self.request
+        toolbar = self.toolbar
+
+        if toolbar._async and 'placeholders[]' in request.GET:
+            # AJAX request to reload page structure.
+            placeholder_ids = request.GET.getlist('placeholders[]')
+            self.placeholders = Placeholder.objects.filter(pk__in=placeholder_ids)
+        else:
+            if toolbar.structure_mode_active and not toolbar.uses_legacy_structure_mode:
+                # User has explicitly requested structure mode.
+                # and the object (page, blog, etc..) allows for the non-legacy structure mode.
+                renderer = toolbar.structure_renderer
+            else:
+                renderer = toolbar.get_content_renderer()
+
+            self.placeholders = renderer.get_rendered_placeholders()
+
+    def add_edit_button(self):
+        if self.page and not user_can_change_page(self.request.user, page=self.page):
+            return
+
+        item = ButtonList(side=self.toolbar.RIGHT)
+        url = get_object_edit_url(self.toolbar.obj, language=self.toolbar.request_language)
+        item.add_button(
+            _('Edit'),
+            url=url,
+            disabled=False,
+            extra_classes=['cms-btn', 'cms-btn-action', 'cms-btn-switch-edit'],
+        )
+        self.toolbar.add_item(item)
+
+    def add_preview_button(self):
+        if self.page and not user_can_change_page(self.request.user, page=self.page):
+            return
+
+        item = ButtonList(side=self.toolbar.RIGHT)
+        url = get_object_preview_url(self.toolbar.obj, language=self.toolbar.request_language)
+        item.add_button(
+            _('Preview'),
+            url=url,
+            disabled=False,
+            extra_classes=['cms-btn', 'cms-btn-switch-save'],
+        )
+        self.toolbar.add_item(item)
+
+    def add_structure_mode(self):
+        if (
+            self.page and not self.page.application_urls and
+            not user_can_change_page(self.request.user, page=self.page)
+        ):
+            return
+
+        if not any(ph for ph in self.placeholders if ph.has_change_permission(self.request.user)):
+            return
+
+        self.add_structure_mode_item()
+
+    def add_structure_mode_item(self, extra_classes=('cms-toolbar-item-cms-mode-switcher',)):
+        structure_active = self.toolbar.structure_mode_active
+        edit_mode_active = (not structure_active and self.toolbar.edit_mode_active)
+
+        with force_language(self.current_lang):
+            if self.request.user.has_perm('cms.use_structure'):
```

Comment:
do this check in the caller. why call method and enter context manager to do a noop